### PR TITLE
vpr: differing modes between lib_nets fix

### PR DESCRIPTION
**Pull request by acomodi**
This PR is to solve https://github.com/SymbiFlow/symbiflow-arch-defs/issues/372. The idea is to retry to route `intra_lib_nets` with the right mode whenever a `Differing modes` error is encountered.

Right now there still is no check whether a net is not routable at all (meaning that all the modes are illegal) and this could end up in an infinite loop. 

It is still to be tested whether parts of the code that are not related to this issue are not affected by the changes of this PR.

Signed-off-by: Alessandro Comodi <acomodi@antmicro.com>

### CODE REVIEW

**Comment by litghost** [comment]
> Right now there still is no check whether a net is not routable at all (meaning that all the modes are illegal) and this could end up in an infinite loop.

Ya, that needs to be fixed before we can merge.

File: vpr/src/pack/cluster_router.cpp
Comment (litghost) [inline vpr/src/pack/cluster_router.cpp]: This is also deeping the stack, probably not a good plan.

Comment (acomodi) [reply vpr/src/pack/cluster_router.cpp]: Yes, I agree, I am not a fan of recursive calls. Now that I know exactly what the issue is I will think of a non-recursive solution

Comment (acomodi) [reply vpr/src/pack/cluster_router.cpp]: Done, even though I believe it should be further refined

**Comment by litghost** [comment]
Seems like you still need to detect if whether all modes are illegal and terminate the search?

**Comment by acomodi** [comment]
> Seems like you still need to detect if whether all modes are illegal and terminate the search?

Yep, I'll push it in a few minutes, I need to check whether it works

**Comment by acomodi** [comment]
@litghost I have added the check which will call `VPR_THROW`. I think it is the most clean solution as not having any more modes at disposal should end up in a fatal error.

**Comment by litghost** [comment]
> > Seems like you still need to detect if whether all modes are illegal and terminate the search?
> 
> Yep, I'll push it in a few minutes, I need to check whether it works

Agreed.  Next step is add some test cases to VPR.  If you need help, ask in #vtr-dev

**Comment by acomodi** [comment]
Build is failing, I will need to check whether this PR introduced a regression

**Comment by acomodi** [comment]
I have checked and confirmed that a regression test is failing with this PR

**Comment by litghost** [comment]
Please squash into one commit.

**Review by litghost — APPROVED** [review]

**Comment by acomodi** [comment]
> Please squash into one commit.

Maybe one commit for regression test and the other with the VPR fix?

**Comment by litghost** [comment]
> > Please squash into one commit.
> 
> Maybe one commit for regression test and the other with the VPR fix?

Sounds good.

**Comment by acomodi** [comment]
Done.

**Comment by litghost** [comment]
I'm going to wait for green CI, then merge.  Thanks!

**Pull request merged**
